src/Http/Middleware/CsrfProtectionMiddleware.php

```diff
@@ -227,9 +240,60 @@ protected function _createToken(): string
      */
     public function createToken(): string
     {
-        $value = Security::randomString(static::TOKEN_VALUE_LENGTH);
+        $value = Security::randomBytes(static::TOKEN_VALUE_LENGTH);
+
+        return base64_encode($value . hash_hmac('sha1', $value, Security::getSalt()));
+    }
+
+    /**
+     * Apply entropy to a CSRF token
+     *
+     * To avoid BREACH apply a random salt value to a token
+     * When the token is compared to the session the token needs
+     * to be unsalted.
+     *
+     * @param string $token The token to salt.
+     * @return string The salted token with the salt appended.
+     */
+    public function saltToken(string $token): string
+    {
+        $decoded = base64_decode($token);
+        $length = strlen($decoded);
+        $salt = Security::randomBytes($length);
+        $salted = '';
+        for ($i = 0; $i < $length; $i++) {
+            // XOR the token and salt together so that we can reverse it later.
+            $salted .= chr(ord($decoded[$i]) ^ ord($salt[$i]));
+        }
+
+        return base64_encode($salted . $salt);
+    }
+
+    /**
+     * Remove the salt from a CSRF token.
+     *
+     * If the token is not TOKEN_VALUE_LENGTH * 2 it is an old
+     * unsalted value that is supported for backwards compatibility.
+     *
+     * @param string $token The token that could be salty.
+     * @return string An unsalted token.
+     */
+    public function unsaltToken(string $token): string
+    {
+        $decoded = base64_decode($token);
+        if ($decoded === false || strlen($decoded) !== static::TOKEN_WITH_CHECKSUM_LENGTH * 2) {
```

Comment:
`base64_decode()` returns false only if 2nd argument is `true` https://www.php.net/manual/en/function.base64-decode.php

Reply:
Good catch. Looks like they have a special case somewhere other than the functionMap I checked. No error when 2nd argument is true.

Reply:
Thanks for assist on this :bow:. I added the strict parameter to all the places we decode tokens from external sources.